scalalib/src/mill/scalalib/ScalaModule.scala

```diff
@@ -20,6 +20,29 @@ trait ScalaModule extends JavaModule with TestModule.ScalaModuleBase { outer =>
   type ScalaModuleTests = ScalaTests
 
   trait ScalaTests extends JavaModuleTests with ScalaModule {
+    try {
+      if (
+        Class.forName("mill.scalajslib.ScalaJSModule").isInstance(outer) && !Class.forName(
+          "mill.scalajslib.ScalaJSModule$ScalaJSTests"
+        ).isInstance(this)
+      ) throw new Exception(
```

Comment:
You can throw a `MillException` or even a `BuildScriptException` here, which will result in a nicer error message in the end.